bucketlist_api/models.py

```diff
@@ -0,0 +1,211 @@
+"""Script defines models for the application."""
+
+import os
+from flask import Flask, request, jsonify, g, url_for
+from sqlalchemy import desc
+from datetime import datetime
+from passlib.apps import custom_app_context as pwd_context
+from bucketlist_api import create_app, db
+from bucketlist_api.config import DevConfig
+from itsdangerous import (TimedJSONWebSignatureSerializer
+                          as Serializer, BadSignature, SignatureExpired)
+
+# initialization
+app = create_app(DevConfig)
+
+
+class BaseModel(db.Model):
+    """The base model for implementing features common to all the models.
+
+    Inherits:
+        db.Model
+
+    Attributes:
+        id: [int] id single model in the database
+        date_created: [datetime] date function was created
+    """
+    __abstract__ = True
+    id = db.Column(db.Integer, primary_key=True)
+    date_created = db.Column(db.DateTime, default=db.func.now())
+
+    def update(self, commit=True, **kwargs):
+        """Update specific fields of a record."""
+        # Prevent changing ID of object
+        kwargs.pop('id', None)
+        for attr, value in kwargs.items():
+            if value is not None:
+                setattr(self, attr, value)
+        return commit and self.save() or self
+
+    def save(self, commit=True):
+        """Save the record."""
+        db.session.add(self)
+        if commit:
+            db.session.commit()
+        return self
+
+    def delete(self, commit=True):
+        """Remove the record from the database."""
+        db.session.delete(self)
+        return commit and db.session.commit()
+
+
+class BucketListModel(BaseModel):
+    """Base class for bucketlist item.
+
+    Inherits:
+        BaseModel
+
+    Attributes:
+        name: [string] the title of the Item
+        date_modified: [datetime] The last modified date
+    """
+    __abstract__ = True
+
+    name = db.Column(db.String(50), nullable=False)
+    date_modified = db.Column(db.DateTime, default=db.func.now())
+
+    def save(self, commit=True):
+        self.date_modified = db.func.now()
+        super(BucketListModel, self).save()
+
+
+class User(BaseModel):
+    """Provides the database Model for Users.
+
+    Inherits:
+        BaseModel
+
+    Attributes:
+        username: [String] username of the user
+        password_hash: [String] Hashed version of the user password
+        bucketlists: [Model Fk] ForeignKey relationship between User and
+                     BucketList
+
+    """
+    __tablename__ = 'users'
+    username = db.Column(db.String(50), nullable=False)
+    password_hash = db.Column(db.String(70))
+    bucketlists = db.relationship('BucketList', backref='users',
```

Comment:
*Nitppick*: This `backref` name you're using is referring to a single user. It shouldn't be pluralized.